fvm/src/machine/limiter.rs

```diff
@@ -0,0 +1,135 @@
+use wasmtime::ResourceLimiter;
+
+use crate::machine::NetworkConfig;
+
+/// Execution level memory tracking and adjustment.
+pub trait ExecMemory
+where
+    Self: Sized,
+{
+    /// Get a snapshot of the total memory required by the modules on the call stack so far.
+    fn curr_exec_memory_bytes(&self) -> usize;
+
+    /// Push a new frame onto the call stack, and keep tallying up the current execution memory,
+    /// then restore it to the current value when the frame is finished.
+    fn with_stack_frame<T, G, F, R>(t: &mut T, g: G, f: F) -> R
+    where
+        G: Fn(&mut T) -> &mut Self,
+        F: FnOnce(&mut T) -> R;
+}
+
+/// Limit resources throughout the whole message execution,
+/// across all Wasm instances.
+pub struct ExecResourceLimiter {
+    /// Maximum bytes that a single Wasm instance can use.
+    max_inst_memory_bytes: usize,
+    /// Maximum bytes that can be used at any point in time during an execution.
+    max_exec_memory_bytes: usize,
+    /// Total bytes desired so far by all the instances including the currently executing instance on the call stack.
+    curr_exec_memory_bytes: usize,
```

Comment:
I used `usize` because [the ResourceLimiter uses it](https://docs.rs/wasmtime/latest/wasmtime/trait.ResourceLimiter.html), but it would be good to introduce at least a newtype for it for sure and then it's easy to change.